File: baekjoon/Java/P15651_N과_M(3).java
/*
backtracking
*/
public class P15651 {
	static int N, M;
	static List<List<Integer>> ans = new ArrayList<>();
	static StringBuilder sb = new StringBuilder();
	public static void main(String[] args) {
		Scanner sc = new Scanner(System.in);
		N = sc.nextInt();
		M = sc.nextInt();

		dfs_(0, new ArrayList<>());

		System.out.print(sb.toString());
	}
	static void dfs_(int depth, List<Integer> lst) {
		if (depth == M) {
			for (int num : lst) {
				sb.append(num).append(" ");
			}
			sb.append("\n");
			return;
		}
		for(int j = 1; j <=N; j++) {
			lst.add(j);
			dfs_(depth + 1, lst);
			lst.remove(lst.size()-1); // 백트래킹
		}
	}
}

File: baekjoon/Python/P15651_N과_M(3).py
import sys

input = sys.stdin.readline

def dfs(depth, lst):
    if depth == M:
        ans.append(lst)
        return

    for j in range(1, N + 1):
        dfs(depth + 1, lst + [j])


N, M = map(int, input().split())
ans = []
dfs(0, [])
for lst in ans:
    print(*lst)
